source/agora/common/Deserializer.d

```diff
@@ -119,6 +119,46 @@ unittest
         Amount(100), [1, 2, 3]));
 }
 
+/// Default upper bound for deserializeLength & co
+public enum DefaultLengthUpperBound = 0x09D0;
+
+/*******************************************************************************
+
+    Deserialize a length, checks if it's within reasonable bound, and return it
+
+    This is a convenience function that calls `deserializeFull`, and checks
+    the result matches a certain bound.
+
+    It is used everywhere an array is deserialized, for example to prevent
+    a DoS if we received crafted binary data that would lead us to allocates
+    large amounts of memory, either exceeding our available memory,
+    or overloading the GC.
+
+    Params:
+        dg         = Delegate to read binary data for deserialization
+        upperBound = The value of the upper bound (inclusive).
+                     Default to 0x09D0, which is a bit less than a page on most
+                     systems, and allow to have allocations taking a full page
+                     (as the GC puts some data there)
+
+    Throws:
+        if the length is > to `upperBound`
+
+    Returns:
+        The deserialized length
+
+*******************************************************************************/
+
+public size_t deserializeLength (
```

Comment:
This still requires the programmer to use this routine when deserializing a length, which they may forget to use. I guess that's why the commit says `It's not a perfect fix..`

What about in the future replacing this with a `deserializeArray` which does all of this internally, and only allowing arrays to be deserialized with such a function?

Reply:
Yes, that's exactly what I'm moving forward. One of the goal is to turn `T.init.deserialize` into `T.fromBinary` (from member function to static factory).